test/FastTests/Client/QueriesWithCustomFunctions.cs

```diff
@@ -2132,6 +2133,125 @@ select new
+        
+        [Fact]
+        public async Task QueryCmpXchgInnerValue(){
+        
+            using (var store = GetDocumentStore())
+            {
+                await store.Operations.SendAsync(new PutCompareExchangeValueOperation<User>("users/1", new User
+                {
+                    Name = "Karmel",
+                    LastName = "Indych"
+                }, 0));
+                var res = await store.Operations.SendAsync(new GetCompareExchangeValueOperation<User>("users/1"));
+                Assert.Equal("Karmel", res.Value.Name);
+                Assert.True(res.Successful);
+                            
+                using (var session = store.OpenSession())
+                {
+                    session.Store(new User { Name = "Jerry", LastName = "Garcia" }, "users/1");
+                    session.SaveChanges();
+                }
+
+                using (var session = store.OpenSession())
+                {
+                    var query = from u in session.Query<User>()
+                        select new
+                        {
+                            u.Name,
+                            UniqueUser = RavenQuery.CmpXchgValue<User>("users/1").Name,
+                        };
+
+                    Assert.Equal("FROM Users as u SELECT { Name : u.Name, UniqueUser : cmpxchg(\"users/1\").Name }", query.ToString());
+
+                    var queryResult = query.ToList();
+
+                    Assert.Equal(1, queryResult.Count);
+                    Assert.Equal("Karmel", queryResult[0].UniqueUser);
+                }
+            }
+        }
+
+        [Fact]
+        public async Task QueryCmpXchg(){
+        
+            using (var store = GetDocumentStore())
+            {
+                await store.Operations.SendAsync(new PutCompareExchangeValueOperation<string>("Tom","Jerry", 0));
+                await store.Operations.SendAsync(new PutCompareExchangeValueOperation<string>("Zeus","Hera", 0));
+                
+                await store.Operations.SendAsync(new PutCompareExchangeValueOperation<string>("users/1","Thunderstorm", 0));
+                await store.Operations.SendAsync(new PutCompareExchangeValueOperation<string>("users/2","Cat", 0));
+                
+                await store.Operations.SendAsync(new PutCompareExchangeValueOperation<string>("Jerry@gmail.com","users/2", 0));
+                await store.Operations.SendAsync(new PutCompareExchangeValueOperation<string>("Zeus@gmail.com","users/1", 0));
+                            
+                using (var session = store.OpenSession())
+                {
+                    session.Store(new User { Name = "Jerry"}, "users/2");
+                    session.Store(new User { Name = "Zeus"}, "users/1");
+                    session.SaveChanges();
+                }
+
+                using (var session = store.OpenSession())
+                {
+                    var rql = "FROM Users WHERE cmpxchg(Name) = \"Hera\"";
```

Comment:
This is actually something that I'm really against supporting. I've tracked how this is implemented, but intuitively, this require you to do a per value filter, which can be _very_ expensive on large collections.
Instead, we want to support queries such as:

`FROM Users WHERE Name = cmpxchg("Hera") or Name = cmgxchg($name)`

Because this way we find the match once, and then use that to query the index.